### PR TITLE
Datetime parsing improvements and other small tweaks

### PR DESCRIPTION
**Pull request by fornarat**
# Overview

First of all, thanks for putting up this project, I find it super useful.
I used it with some wild CSV files and encounter some issues that I solved along the way, plus some small optimisations, so apologies in advance for pushing so many changes at once, I just had to make it work for my needs ASAP.
Hopefully (much) of this changes make sense to you, I'm willing to elaborate/improve otherwise, as I'd be using this  project intensively.

---

Please preserve this line to notify @iSnow (lead of this repository)

### CODE REVIEW

**Comment by iSnow** [comment]
Hi @fornarat thanks a ton for such an extensive PR. Please have some patience, it is not forgotten and I will review it, but I first have to merge another PR, so it's gonna be another day or two.

File: src/main/java/io/frictionlessdata/tableschema/schema/TypeInferrer.java [hunk 1]
```diff
@@ -36,18 +39,22 @@
     // The order in which the types will be attempted to be inferred.
     // Once a type is successfully inferred, we do not bother with the remaining types.
     private static final List<String[]> TYPE_INFERRAL_ORDER_LIST = new ArrayList<>(Arrays.asList(
-        new String[]{Field.FIELD_TYPE_GEOPOINT, Field.FIELD_FORMAT_DEFAULT},
-        new String[]{Field.FIELD_TYPE_GEOPOINT, Field.FIELD_FORMAT_ARRAY},
-        new String[]{Field.FIELD_TYPE_GEOPOINT, Field.FIELD_FORMAT_OBJECT},
-        new String[]{Field.FIELD_TYPE_DURATION, Field.FIELD_FORMAT_DEFAULT}, // No different formats, just use default.
+        new String[]{Field.FIELD_TYPE_BOOLEAN, Field.FIELD_FORMAT_DEFAULT}, // No different formats, just use default.
```
Comment (iSnow) [inline src/main/java/io/frictionlessdata/tableschema/schema/TypeInferrer.java]: Could you please explain why the change in the ordering of the inferral order list? Which cases would be parsed wrong?

File: src/main/java/io/frictionlessdata/tableschema/field/YearField.java [hunk 2]
```diff
@@ -30,7 +32,10 @@ public YearField(String name, String format, String title, String description,
 
     @Override
     public Year parseValue(String value, String format, Map<String, Object> options) throws InvalidCastException, ConstraintsException {
-        Pattern pattern = Pattern.compile(REGEX_YEAR);
+        if(value.length()>4){
```
Comment (iSnow) [inline src/main/java/io/frictionlessdata/tableschema/field/YearField.java]: I believe this check is not needed, as the `pattern.matcher` will throw anyways.

I whipped up a small test:
     @Test
     @DisplayName("format year")
     void formatNumberField() {
         YearField field = new YearField("test");
         Year val = field.parseValue("12345", null, null);
     }

And yes, it throws. Is there something I am missing or is this check just a precaution?

File: src/main/java/io/frictionlessdata/tableschema/field/DateField.java [hunk 3]
```diff
@@ -33,23 +33,22 @@ public DateField(String name, String format, String title, String description,
     public LocalDate parseValue(String value, String format, Map<String, Object> options)
             throws InvalidCastException, ConstraintsException {
 
-        Pattern pattern = Pattern.compile(REGEX_DATE);
-        Matcher matcher = pattern.matcher(value);
+        if(value.length()>patternFormat.length() || value.length() < "y-M-d".length()){
```
Comment (iSnow) [inline src/main/java/io/frictionlessdata/tableschema/field/DateField.java]: Like for the YearField, I believe this check is superfluous. The matcher would throw anyways for a String like `9-1-2`, right?

File: src/main/java/io/frictionlessdata/tableschema/field/YearField.java (hunk 2 above)
Comment (fornarat) [reply src/main/java/io/frictionlessdata/tableschema/field/YearField.java]: It was intended as a precondition to avoid calling the `pattern.matcher` which comes with some computational overhead, but I'll take it away since the gains are negligible against how it affects readability.

File: src/main/java/io/frictionlessdata/tableschema/field/DateField.java (hunk 3 above)
Comment (fornarat) [reply src/main/java/io/frictionlessdata/tableschema/field/DateField.java]: same as for YearField, I'll roll it back

File: src/main/java/io/frictionlessdata/tableschema/schema/TypeInferrer.java (hunk 1 above)
Comment (fornarat) [reply src/main/java/io/frictionlessdata/tableschema/schema/TypeInferrer.java]: As in the other cases, there is nothing wrong with the current implementation and the change in the ordering shouldn't affect the results. 
The aim was to improve the performance of the parser handling first the cases that I found more often. Although I'm handling an amount of files less than representative, I found that the time-related fields were much more common than location-related fields, so I prioritized one group over the other.
Let's roll it back since it's a few lines of code and this PR was extensive enough and we can discuss this later on a new PR (hopefully with tests that hold my claims)
